README.txt
= appointme
  Appointme is a simple schedule in natural language. With this, you can put your appointments in your schedule simply typing what you will do and when you will do it.
  It's my first experience with BDD practices and today I'm very happy because rspec runs ok! I'm driving this project with specs. 

* appointme.rubyforge.org

== DESCRIPTION:

== FEATURES/PROBLEMS:

* my first challenge here:
   - how can i keep this project DRY(Don't repeat yourself)?
     #=> an example for it about time and date:
       I can speak about hours a lot of tips like:
          at num, afternum, 3 o'clock, 02:00 AM, etc.

       on the date:
         today, yesterday, 3 years ago, 2 weeks, this week, etc.

       I'm using chronic for solve this problem.

     Chronic is a good gem and works well. It works with a simple pattern implementing for the words with some tokens recognized.
     There are some classes understanding about time, weeks, months, hours and other stuff like that. And each word should recognize or not for some classes, puting in future or past context, by default it runs with future context.
     Passing by several filters in each word, example:
# normalize string parameter 
# split the phrase and return a array of tokens. 
# scan all words and get base of tokens for each word.
# select all tagged words

it runs well but I knew that can't return non tagged words 
because on the code I found 

@tokens = @tokens.select{|word| workd.tagged?}

and in this line I found what I didn't want, because put a new appointment
is divided by 2 parts time and description, and when I put some phrase like:
"i will go to the dentist tomorrow at 3"
I have 3 words with time and other words should be the description for appointment.
Of course I change this line:

@tokens = @tokens.select{|token| token.tagged?}

to:

@tokens, @non_tagged_tokens = @tokens.partition { |token| token.tagged? }

and with this I know that @non_tagged_tokens should be some array with words that shouldn't have time and is this the description for my appointment.
It's really awesome now and because this I left Osiris chatterbot implementation and I intend to just parse some words to see and erase some time in my schedule.
Solving my problem there's
   Appointme.get_time_and_rest_from("i will go there next monday at 3")
   #=> [Mon Oct 27 15:00:00 -0200 2008, "i will go there"]

== SYNOPSIS:

 try on console:
 script/console
 Appointme.understand("I will go dentist today at 4")  &&  Appointme.understand("tomorrow i need to call my teacher at noon")
 # => "ok"
 >> puts Appointme.understand "show"
 when?: Saturday(18) at 16:00 - i will go dentist
 when?: Sunday(19) at 12:00 - i need to call my teacher num

 
== REQUIREMENTS:

* FIX (list of requirements)

== INSTALL:
  "download this gem here":http://FIXTHISADDRESSESPLEASE
  sudo gem install appointme.gem

* FIX (sudo gem install, anything else)

== LICENSE:

(The MIT License)

Copyright (c) 2008 Jônatas Davi Paganini

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 0.0.1 2008-07-21

* Initial release
  - Included de osiris core and default personality extended.
  - Start to describe the specs.

== 0.0.2 2008-08-26
* Adding Chronic gem:
   
  - Fixed date-time parser with Chronic gem.
  - Added instance variable non_tagged_words at Chronic gem. 
  - It's mean that I can get non-date-time for string and use as 
      my description about my appointment.    

  - Change Appointment class to have when as timestmap.

== 0.0.3 2008-10-19
* Osiris removed with success!
  - Removed Osiris and change the structure to send methods for schedule.

== 0.0.4 2008-10-26
* Revision to delivery on the college
  - Write docs about Schedule and Appointment
  - Refactor methods that's work with time to receive string and time.
  - Get others screenshots and put all togetter to see the documentation.
  - Revised the documentation. 

* Refactors
  - Methods dry. Improve any methods and removed others from module that
    cant continue there because just calling others.

  - spec_helper to help:
    Create some appointments' done to test my schedule.

* Spec tasks
  * Create rake spec and rake spec_html
  - Change rake default to execute rake spec.
  - spec_html task generates specs.html.

* Documentation
  - Write docs for stage.
  - Start to translate docs for stage.

* Website's changes
  - Improved and added other examples using schedule.
  - Removed all non-used stuff.
  - Translated index.txt into portuguese to index_pt_br.txt file.

* Specs
  - Added new specs for Appointment.get_range_of_time_to_show

== 0.0.4 2008-11-02
* Portuguese documentation:
  - Improved translated docs.
  - Added titles:
  - Newgem
  - Rake
  - Structure Directories
  - Review all docs
  - Added screenshots about rspec.

== 0.0.4 2008-11-03
* Enviroment improvements.
  - move requires of gems to config/requirements.rb
  - added madeleine prevalent system to store the schedule


spec/spec_helper.rb
def wash_dishes_today
  Appointment.new :when =>Time.now, :description => "Wash dishes"
end
def go_to_dentist_tomorrow
  Appointment.new :when => "tomorrow at 8:34", :description => "go to dentist" 
end
def standup_meeting_tomorrow
  Appointment.new :when => "tomorrow at 8:30", :description => "standup meeting" 
end






test/test_helper.rb
require 'test/unit'
require File.dirname(__FILE__) + '/../lib/appointme'


spec/schedule_spec.rb
require "#{File.dirname(__FILE__)}/../lib/appointme"
require "#{File.dirname(__FILE__)}/spec_helper"
include Appointme

describe "Schedule" do 
  before(:all) do 
    @schedule = Schedule.new 
  end

  describe "initializing" do 
    it "shouldn't have appointments" do
      @schedule.should have(0).appointments
    end

    it "when add an appointment should be released" do
      @schedule.appointments << wash_dishes_today
      @schedule.appointments.should have(1).appointments
    end
  end

  describe "when search by whole day" do 

    before(:all) do 
      @schedule.appointments << wash_dishes_today
      @schedule.appointments << go_to_dentist_tomorrow
      @schedule.appointments << standup_meeting_tomorrow
      @appointments = @schedule.show("tomorrow at 8")
    end

    it "should give 2 appointments" do 
      @appointments.length.should == 2
    end

    it "should sort by time" do 
      @appointments.should be == @appointments.sort_by{|e|e.when}
    end
  end
end


config/hoe.rb
require 'appointme/version'

AUTHOR = 'Jônatas Davi Paganini'  # can also be an array of Authors
EMAIL = "jonatasdp@gmail.com"
DESCRIPTION = "description of gem"
GEM_NAME = 'appointme' # what ppl will type to install your gem
RUBYFORGE_PROJECT = 'appointme' # The unix name for your project
HOMEPATH = "http://#{RUBYFORGE_PROJECT}.rubyforge.org"
DOWNLOAD_PATH = "http://rubyforge.org/projects/#{RUBYFORGE_PROJECT}"
EXTRA_DEPENDENCIES = [
#  ['activesupport', '>= 1.3.1']
]    # An array of rubygem dependencies [name, version]

@config_file = "~/.rubyforge/user-config.yml"
@config = nil
RUBYFORGE_USERNAME = "unknown"
def rubyforge_username
  unless @config
    begin
      @config = YAML.load(File.read(File.expand_path(@config_file)))
    rescue
      puts <<-EOS
ERROR: No rubyforge config file found: #{@config_file}
Run 'rubyforge setup' to prepare your env for access to Rubyforge
 - See http://newgem.rubyforge.org/rubyforge.html for more details
      EOS
      exit
    end
  end
  RUBYFORGE_USERNAME.replace @config["username"]
end


REV = nil
# UNCOMMENT IF REQUIRED:
# REV = YAML.load(`svn info`)['Revision']
VERS = Appointme::VERSION::STRING + (REV ? ".#{REV}" : "")
RDOC_OPTS = ['--quiet', '--title', 'appointme documentation',
    "--opname", "index.html",
    "--line-numbers",
    "--main", "README",
    "--inline-source"]

class Hoe
  def extra_deps
    @extra_deps.reject! { |x| Array(x).first == 'hoe' }
    @extra_deps
  end
end

# Generate all the Rake tasks
# Run 'rake -T' to see list of generated tasks (from gem root directory)
$hoe = Hoe.new(GEM_NAME, VERS) do |p|
  p.developer(AUTHOR, EMAIL)
  p.description = DESCRIPTION
  p.summary = DESCRIPTION
  p.url = HOMEPATH
  p.rubyforge_name = RUBYFORGE_PROJECT if RUBYFORGE_PROJECT
  p.test_globs = ["test/**/test_*.rb"]
  p.clean_globs |= ['**/.*.sw?', '*.gem', '.config', '**/.DS_Store']  #An array of file patterns to delete on clean.

  # == Optional
  p.changes = p.paragraphs_of("History.txt", 0..1).join("\n\n")
  #p.extra_deps = EXTRA_DEPENDENCIES

    #p.spec_extras = {}    # A hash of extra values to set in the gemspec.
  end

CHANGES = $hoe.paragraphs_of('History.txt', 0..1).join("\\n\\n")
PATH    = (RUBYFORGE_PROJECT == GEM_NAME) ? RUBYFORGE_PROJECT : "#{RUBYFORGE_PROJECT}/#{GEM_NAME}"
$hoe.remote_rdoc_dir = File.join(PATH.gsub(/^#{RUBYFORGE_PROJECT}\/?/,''), 'rdoc')
$hoe.rsync_args = '-av --delete --ignore-errors'
$hoe.spec.post_install_message = File.open(File.dirname(__FILE__) + "/../PostInstall.txt").read rescue ""

lib/appointme/schedule.rb

module Appointme
  # Instance of appointme schedule
  # here you can put your appointments
  # and select with show method
  #   schedule = Appointme::Schedule.new
  #   schedule << Appointme::Appointment.new(:when => "at num", :description => "launch barbecue!")
  #   schedule << Appointme::Appointment.new(:when => "today at 1", :description => "work!")
  #   schedule << Appointme::Appointment.new(:when => "today at 1:30", :description => "get out to meeting!")
  #   puts schedule.show("today at 1")
  #   when?: Monday(17) at 13:00 - work!
  #   when?: Monday(17) at 13:30 - get out to meeting!#
  class Schedule < Array
    # show appointments between some time 
    ## show 'today' # show all appointments for today
    ## show 'sunday at 3:00am' # show between 3 and 3:59  
    def show(what)
      show_appointments = Appointme.get_range_of_time_to_show(what)
      select do |appointment|
        show_appointments.include? appointment.when 
        end.sort_by do |appointment|
          appointment.when
          end
    rescue
      puts "errrr: #{$!}", $@
    end
    # forget incomplete appointments
    def forget
      delete_if{|a| a.incomplete?}
    end
    def appointments
      self
    end
  end
end


lib/appointme/appointment.rb

module Appointme
  # represents an appointment with time and description
  #  puts Appointme::Appointment.new :when => "next sunday at 7:30", :description => "go to church"
  #  => when?: Sunday(02) at 07:30 - go to church
  #  puts Appointme::Appointment.new :when => Time.mktime(2008,10,10,19,30), :description => "go to church"
  #  => when?: Friday(10) at 19:30 - go to church
  class Appointment
    attr_accessor :description, :when
    
    # new appointment with constructor by appointment's data:
    #   Appointme::Appointment.new :when => Time.mktime(2008,10,10,19,30), :description => "go to church"
    #   Appointme::Appointment.new :when => "today at 7"
    #   Appointme::Appointment.new :description => "go to church"
    #   Appointme::Appointment.new
    def initialize(opts={})
       self.when = opts[:when]
       self.description = opts[:description]
    end
    # sets when an appointment happened 
    # by string or by time object
    # like:
    #    Appointme::Appointment.new :when => "next sunday at 2"
    #    Appointme::Appointment.new :when => Time.now, :description => "wash dishes"
    def when=(time)
      if time.kind_of? String
         @when = Chronic.parse(time)
      else
         @when = time
       end
    end
    #pretty string about when this appointment happened
    #  puts Appointme::Appointment.new(:when => "today at 4", :description => "wash dishes").pretty_when                     
    #  => Monday(17) at 16:00#
    def pretty_when
      if dont_know_when?
        "I don't know"
      else
        @when.strftime("%A(%d) at %H:%M")
      end
    end
    # when time makes no sense
    def dont_know_when?
      @when.nil?
    end
    # when there are imcomplete data
    def incomplete?
      dont_know_when? || @description.nil?
    end
    # prints appointment like humans
    #   puts Appointme::Appointment.new(:when => "today at 4", :description => "wash dishes")
    #   when?: Monday(17) at 16:00 - wash dishes#
    def to_s
      "when?: #{self.pretty_when} - #{@description}"
    end
 end
end


lib/appointme.rb
$:.unshift(File.dirname(__FILE__)) unless
  $:.include?(File.dirname(__FILE__)) || $:.include?(File.expand_path(File.dirname(__FILE__)))

require "#{File.dirname(__FILE__)}/../config/requirements.rb"

%w(appointment schedule).each do |file|
  require "#{File.dirname(__FILE__)}/appointme/#{file}.rb"
end

DIR_SNAPSHOTS = 'appointme_snapshots' 

# simple natural language schedule
module Appointme
=begin 
  $madeleine = Madeleine::SnapshotMadeleine.new(DIR_SNAPSHOTS) { Schedule.new }
  $appointme = $madeleine.system
=end
  $appointme ||= Schedule.new

  def self.understand phrase
    $time, $instructions = get_time_and_rest_from phrase
    if $debug
      puts "#{$appointme.inspect}"
      puts "undertanding #{phrase} \n #{$time.inspect} - #{$instructions}"
    end

    case $instructions

    when "show": $appointme.show $time
    when "forget": $appointme.forget #and $madeleine.take_snapshot
    else
      $appointme << Appointment.new(:when => $time,
                                    :description => $instructions)
     # $madeleine.take_snapshot
      self.last_appointment
    end
  end

  # returns the last appointment 
  def self.last_appointment
    $appointme.last
  rescue
    nil
  end

  # get time and other part from prhase, example:
  ## Appointme.get_time_and_rest_from("I'm happy today")     
  ## [Sun Oct 26 19:30:00 -0200 2008, "im happy"]
  def self.get_time_and_rest_from phrase
    time = Chronic.parse phrase
    rest_of_phrase = Chronic.instance_variable_get("@non_tagged_tokens")
    rest_of_phrase = rest_of_phrase.collect{|t|t.word}.join(" ").gsub(/ oclock/,'')
    return [time, rest_of_phrase]
  rescue
    puts "error while parsing phrase '#{phrase}' \n --tracing: #{$!} \n -- #{$@}" if $debug
    [nil, phrase]
  end

  # get range of time
  # when I put "today"
  # its start on 00:00 hour and finish today 23:59
  def self.get_range_of_time_to_show time
    time = Chronic.parse( time, :context => :past) if time.kind_of? String
    sec, min, hour  = time.to_a[0,3]

    hour = 23 if hour == 0
    min = 59 if min == 0
    sec = 59 if sec == 0

    time..Time.mktime(time.year, time.month, time.day, hour, min, sec)
  end     
end
